tests/trainer/test_supporters.py

```diff
@@ -182,3 +191,73 @@ def test_combined_loader_sequence_max_size_cycle():
         assert len(item) == 2
 
     assert idx == len(combined_loader) - 1
+
+
+class IndexedRandomDataset(RandomDataset):
+
+    def __getitem__(self, index):
+        return {"index": index, "batch": self.data[index]}
+
+
+class TestModel(BoringModel):
+
+    def __init__(self, numbers_test_dataloaders, save_preds_on_dl_idx):
+        super().__init__()
+        self._numbers_test_dataloaders = numbers_test_dataloaders
+        self._save_preds_on_dl_idx = save_preds_on_dl_idx
+
+    def test_step(self, batch, batch_idx, dataloader_idx=None):
+        indexes, x = batch["index"], batch["batch"]
+        output = self.layer(x)
+        loss = self.loss(batch, output)
+
+        if dataloader_idx is not None and dataloader_idx == 1 and not self._save_preds_on_dl_idx:
+            return {"y": loss}
+
+        predictions = []
+        for idx in range(len(indexes)):
+            id = indexes[idx]
+            prediction = output[idx]
+            predictions.append({"id": id, "prediction":prediction})
+
+        self.add_predictions(predictions)
+        return {"y": loss}
+
+    def create_dataset(self):
+        return torch.utils.data.DataLoader(IndexedRandomDataset(32, 64), batch_size=2)
+
+    def test_dataloader(self):
+        return [self.create_dataset() for _ in range(self._numbers_test_dataloaders)]
+
+
+@pytest.mark.skipif(not os.getenv("PL_RUNNING_SPECIAL_TESTS", '0') == '1',
+                    reason="test should be run outside of pytest")
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
+@pytest.mark.parametrize('save_preds_on_dl_idx', [False, True])
+@pytest.mark.parametrize('accelerator', ["ddp"])
+@pytest.mark.parametrize('gpus', [1, 2])
+@pytest.mark.parametrize('dl_idx', [1, 2])
+def test_prediction_collection_ddp(tmpdir, save_preds_on_dl_idx, accelerator, gpus, dl_idx):
+
+    """
+    Test `PredictionCollection` reduce properly in ddp mode
+    """
+    model = TestModel(dl_idx, save_preds_on_dl_idx)
+    model.test_epoch_end = None
+    limit_test_batches = 2
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_test_batches=limit_test_batches,
+        accelerator=accelerator,
+        gpus=gpus
+    )
+    results: List = trainer.test(model)
+    assert len(results) == dl_idx
+    for dl_idx in range(dl_idx):
+        result = results[dl_idx]
+        if not save_preds_on_dl_idx and dl_idx == 2 and dl_idx == 1:
+            assert "predictions" not in result
+        else:
+            assert "predictions" in result
```

Comment:
The right line was `if not save_preds_on_dl_idx and num_dl_idx == 2 and dl_idx == 1:`. When evaluating for dataloader_idx == 1 and save_preds_on_dl_idx = False, it will skip `add_predictions`. Therefore, it ensures predictions isn't contained in result object.